### ISSUE TITLE
Can't create a AppGw rule when using a private IP address

### ISSUE THREAD

**Issue by timvanderweijde** [Azure/azure-cli#16086]
## Describe the bug
Can't create a rule when using a private IP address.

**Command Name**
`az network application-gateway rule create`

**Errors:**
```
BadRequestError: FrontendIpConfiguration /subscriptions/<someguid>/resourceGroups/some-resource-name/providers/Microsoft.Network/applicationGateways/some-gateway-name/frontendIPConfigurations/appGatewayPrivateFrontendIP cannot use private IP address <someIPhere> as it is either allocated to a resource or being cleaned up.
```

## To Reproduce:
Gateway, backendpool, httpsettings, http listeners are created. Http Listerens Frontend IP = Private.

- `az network application-gateway rule create -g {} --gateway-name {} -n {} --address-pool {} --http-listener {} --http-settings {} --rule-type {} --rewrite-rule-set {}`

## Expected Behavior
Just created a rule.

## Environment Summary
```
Windows-10-10.0.19041-SP0
Python 3.6.8
Installer: MSI

azure-cli 2.14.0 *

```

**Comment by timvanderweijde** [comment]
Extra information: sometimes it is being created correctly, sometimes it doesn't.

We have a create Gateway script. We deleted the application gateway and run the script again with success. But...we did it again, and the error was again there.

It sounds like a timing issue...however, we don't use the -no-wait.

**Comment by yonzhan** [comment]
@haroldrandom for awareness.

**Comment by haroldrandom** [comment]
@timvanderweijde Could you try 2.15.0? I patched a similar [fix ](https://github.com/Azure/azure-cli/pull/15757) like this one in 2.15.0

**Comment by yonzhan** [comment]
@timvanderweijde any update?

**Comment by timvanderweijde** [comment]
@yonzhan still the same problem. It cannot create the AppGW rule.

We tried to wait for 1 hour, 3 hours and last full weekend without sucess.

**Comment by haroldrandom** [comment]
@timvanderweijde Is the error still the same?

**Comment by wapatmnl** [comment]
Hi, I am taking over this issue from my colleague Tim. 

The issue is still present, also using CLI 2.15.1.
We currently have an instance running where we the rule creation still fails this way.
Error is still the same: cannot use private IP address <ip> as it is either allocated to a resource or being cleaned up.

As it blocks progress in one of our projects I am also happy to create a support case, as advised by another MS contact.

Kind regards,

Walter

**Comment by wapatmnl** [comment]
Hi,

I started to put in the support request when the support wizard popped up with some additional information that better explains what is going on behind the curtains. There are the typical instruction to remove the frontendIP configuration and pick another address, which do not really explain how to avoid the issues. But in particular a phrase like "...enter a non-conflicting IP address, preferably from the end of the subnet address space..." describes what is going on, as also does "Currently when you configure a private frontend IP address without associating it to a listener and a rule, the IP is not reserved in the subnet, so the same IP address getting assigned to an instance of the Application Gateway."

In our setup scripts the AG is created and all the other items are bolted on afterwards. Now when the gateway then directly starts, it allocates the first IPs in the subnet, among which is the one we reserved for our frontendIp. If that happens fast enough, then our following statements for creating the rules will fail. And when you go down this rabbit hole, you literally never get it back, even stopping the AG does not free up the IP (at least not directly). If you are lucky and your script directly runs without flaws and fast enough, you might beat the startup and get the IP reserved before the AG itself claims it.

This indeed is solved by just picking an IP that is unlikely to be picked by the AG, for example at the end of the subnet. That does mean for us that we have to adjust a set of planned DNS entries etc. (further down the project that would have led to firewall adjustments, NSGs, etc.). And in addition, we might need to resize the subnet for capacity in the long run, which also makes an IP from "the end" more complex.

There is just a kind of randomness to the whole setup that does not fit will with carefully laid out plans. It would be much nicer to break the dependency between the selected private listener IP and the subnet range chosen for the AGs instance usage. For example with an explicit reservation configuration field or something like that. 
And as an additional hint: one does find the answer in the end, but that could have been coming up a lot sooner in terms of documentation...

Anyway, thanks for the support, we can continue now.

Kind regards,

W.
